pallene/ir.lua

```diff
@@ -45,12 +44,17 @@ function ir.VarDecl(name, typ)
 end
 
 function ir.Function(loc, name, typ)
+    -- We need to store the number of upvalues separately (instead of just using #captured_vars) because
+    -- even though the captured_vars table is append-only, the actual number of upvalues may change
+    -- after the constant propagation pass. We cannot mutate the `captured_vars` table to reflect this change
+    -- because we still need type information of the propagated upvalues (which become constant values).
```

Comment:
This is awkward. Do you have an idea of what we would need to do to avoid this problem?

Not necessarily for this PR. I'm OK with keeping this workaround for now.

Reply:
Another idea I tried was to get rid of the entries in `captured_vars` that get constant propagated, thereby shortening it.
However, this way we lose not only the decl but also the type information for the upvalue, which is necessary in `coder.lua`.

So one way could to be to use a separate map for `typ_of_upvalue` , and then replace `captured_vars` with an updated version that has the correct number of upvalues in the constant propagation pass.

Reply:
Could you give an example of how we might need that type information in the code? I though that if a variable or upvalue is constant propagated, then all it's uses should have been removed.

Reply:
Turns out the reason was some of my experiments with the coder that I didn't commit xD
I git reset and readjust the captured_vars and now it works.

Reply:
Great! One less thing for us to worry about later.